added cache_dir and cache_workers options to cache results

gpakosz opened the pull request:
:boom:

implemented #83

toy commented:
Hi Gregory, sorry for the wait and thank you for contribution!
I've looked through commit, but did not yet have time to properly check it and I'll try to find time for that soon.

gpakosz commented:
Hi, I rebased the branch onto the master branch.

toy commented on lib/image_optim.rb:
```diff
@@ -93,15 +104,21 @@ def workers_for_image(path)
-    ImagePath::Optimized.new(result, original)
+    return unless optimized
+
+    @cache.write(optimized, original)
```
I would switch the order of the arguments as original works as a key

toy commented on lib/image_optim.rb:
```diff
@@ -93,15 +104,21 @@ def workers_for_image(path)
-    result = Handler.for(original) do |handler|
+
+    cached = @cache.read(original)
+    return ImagePath::Optimized.new(cached, original) unless cached.nil?
```
It may be shorter if you rename cached to `optimized` and run handler and write cache if nil is returned

toy commented on lib/image_optim/bin_resolver/bin.rb:
```diff
@@ -11,11 +12,12 @@ class Bin
         @name = name.to_sym
         @path = path.to_s
         @version = detect_version
+        @digest = File.exist?(@path) && Digest::SHA1.file(@path).hexdigest
```
Can be a method with caching using `defined?(@digest)`

toy replied:
I am also think if the sha1 is really needed or just the version will be enough

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,69 @@
+      return unless (@cache_dir = image_optim.cache_dir)
+      @cache_workers = image_optim.cache_workers
+
+      FileUtils.mkdir_p(@cache_dir)
```
Using `FSPath` for paths is good ;)

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,69 @@
+      @options_by_format = Hash[workers_by_format.map do |k, v|
+        [k, Marshal.dump(v.map(&:inspect))]
+      end]
+      @workers_by_format = Hash[workers_by_format.map do |k, v|
```
Name `@workers_by_format` is confusing as inside are bins and not workers

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,69 @@
+      FileUtils.mkdir_p(@cache_dir)
+
+      @options_by_format = Hash[workers_by_format.map do |k, v|
+        [k, Marshal.dump(v.map(&:inspect))]
```
Marshal is not really needed, you can just join strings

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,69 @@
+        [k, Marshal.dump(v.map(&:inspect))]
+      end]
+      @workers_by_format = Hash[workers_by_format.map do |k, v|
+        [k, Marshal.dump(
```
Same comment about Marshal

toy commented:
Hi Gregory, I am sorry for such a long wait.

Few things to discuss:
I think it is better to create one cache file and not 2 or 3, otherwise it is not atomic, there are too many files and if running with different options the cache will be overwritten.
The name can be combined or concatenated from options and binaries digests.
Also I think it is better to use sub directories as in `.git/objects` instead of putting everything in one directory.

gpakosz commented:
I followed every single remark you made except using `FSPath` because it didn't feel like an improvement to use it just to join paths and I need `FileUtils.copy_file` anyways.

I made a separate commit. Once reviewed I suggest I squash the two commit before force pushing the branch one last time.

gpakosz commented:
Well CI has errors unrelated to my changes, can you have a look?

    Installing rack 1.6.4
    An error occurred while installing mime-types-data (3.2016.0221), and Bundler

gpakosz commented:
Hi Ivan, when do you think we can roll another round of review for this PR?

toy commented:
Hi Gregory, thanks for poking me about this.
I've already looked at it and checked how it works, but did not manage to find time to write everything. Points to discuss:

1. Currently an already optimised file will not be cached as `write` will return if `optimized` is not present. Simple solution would be to just write the original file to cache, but it looks like wasting disk space. Two other solutions if using the same path are: to create an empty file (then atomic creation of cache is a must) or create a symlink to self. What do you think?

2. Is binary sha1 really better or the bin version is enough?

3. There is no need for `Marshal`, as strings can be just joined together.

4. Why FSPath:
```ruby
cached = @cache_dir / digest(original)
return unless cached.exist?
```
instead of
```ruby
cached = File.join(@cache_dir, digest(original))
return unless File.exist?(cached)
```
Also `FileUtils.copy_file` is not atomic. A method similar to [`ImagePath#replace`](https://github.com/toy/image_optim/blob/master/lib/image_optim/image_path.rb#L41), which would create a temporary file in cache directory, copy optimized file over it and rename it to cache path would make it atomic.

gpakosz commented:
Hi Ivan,

Here are my answers:

1. I think I would go for a symlink

2. Using the binary SHA1 seems more robust in case a binary fails to update its version information properly. Are you concerned about performance when computing SHA1s?

3. `Marshal` seems like a robust way of dumping a state to a string. Again are you concerned about the performance?

4. I'll look into `FSPath`

toy commented:
1. Agreed.
2. I was thinking more about two bins of same version compiled differently. So to me it seems like deciding between been a bit too precise and a bit not enough precise.
3. I'm concerned about marshal version, but also has not a lot of purpose when there are already just strings. It is also important to ensure that the final strings to digest can't be different for identical options/bins due to order.

gpakosz commented:
I implemented the changed discussed so far.

Note that the symlink only serves the purpose of going through workers again, but `optimize_image` will still return `nil`. This is because I don't think it's a good idea to have `optimize_image` behavior being affected by whether or not caching is enabled.

toy commented:
Great! I'll do my best not to delay review.
And sure, the cache should not change behaviour.

toy commented on lib/image_optim.rb:
```diff
@@ -93,15 +104,22 @@ def workers_for_image(path)
-    ImagePath::Optimized.new(result, original)
+
+    return unless optimized
+    return unless File.file?(optimized) && !File.symlink?(optimized)
```
This should be handled inside `Cache` class. I can propose to combine `Cache#read` and `Cache#write` to a method accepting block, so if there is cache hit, it will be returned, if there is a symlink, nil will be returned, otherwise the block will run and result will be written in cache. So the call would be like:
```ruby
optimized = @cache.fetch(original) do
  Handler.for(original) do |handler|
    ...
  end
end
```

toy commented on lib/image_optim.rb:
```diff
@@ -37,6 +38,12 @@ class ImageOptim
+  attr_reader :cache_dir
+
+  # Cache workers digest
+  attr_reader :cache_workers
```
`cache_workers` name seems not descriptive. I did not find some obvious name, but something like `cache_worker_digests` or `cache_by_digest`, what do you think?

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,68 @@
+      return unless image_optim.cache_dir
+      @cache_dir = FSPath.new(image_optim.cache_dir)
+      @cache_workers = image_optim.cache_workers
+      @options_by_format = Hash[workers_by_format.map do |k, v|
```
Descriptive parameter names, like `format, workers`?

gpakosz replied:
I don't get it

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,68 @@
+      return unless @cache_dir
+
+      cached = @cache_dir / digest(original)
+      return unless File.exist?(cached)
```
You don't use the power of `Pathname` – `cached.exist?`

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,68 @@
+      @options_by_format = Hash[workers_by_format.map do |k, v|
+        [k, v.map(&:inspect).sort.join(', ')]
+      end]
+      @bins_by_format = Hash[workers_by_format.map do |k, v|
```
Same comment about parameter names

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,68 @@
+
+      digest = digest(original)
+      cached = @cache_dir / digest
+      FileUtils.mkdir_p(File.dirname(cached))
```
`cached.dirname.mkpath`

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,68 @@
+
+      if optimized
+        t = Time.now.strftime('%Y%m%d')
+        tmp = @cache_dir / "#{digest}#{t}-#{$PID}-#{rand(0x100000000).to_s(36)}"
```
While this name will be most likely unique, I think it would still be better to use `Tempfile` through `FSPath` like:
```ruby
FSPath.temp_file_path(something, @cache_dir) do |tmp|
  # move and rename
end
```
`something` could have been just digest, but it already contains slash, which will create a problem for `Tempfile`

gpakosz replied:
But doing so, under the hood `Tempfile` opens a file handle right? In which case that's useless and a waste imho.

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,68 @@
+        FileUtils.mv(optimized, tmp, :force => true)
+        File.rename(tmp, cached)
+      else
+        FileUtils.ln_sf(original, cached)
```
Original may disappear and the linked file is not used anyway, so it does not make sense to create a link to it, but maybe a fake link to '-' or basename of cached.

toy commented:
Please also add description of two new options to [Options section of README.markdown](../README.markdown#options) and to [`OptionParser`](../lib/image_optim/runner/option_parser.rb), so they are exposed to command line

gpakosz commented:
I implemented what's been discussed except `FSPath.temp_file_path` because it seems wasteful to open file handles just for the sake of having a unique name

gpakosz commented:
I forgot to document the options properly, I'll do that later along with adding few tests to make sure `cache_dir` is a directory if it already exists

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,71 @@
+    end
+
+    def fetch(original)
+      return block_given? ? yield : nil if @cache_dir.nil?
```
`fetch` should not be used without block, so I think you should call `yield` without checking `block_given?`

gpakosz commented:
Updated, and I reworked the tests. By now, since cache file digest is computed by updating file SHA1 with options and optionally worker digests, there's no real distinction between "cached file doesn't exist" and "options and/or worker digests don't match".

toy commented:
About temporary file – it is an extra file handle (even worse for old ruby version as it was creating a locking directory), but also the handle can be used with `FSPath.temp_file` to write to it instead of using `FileUtils.mv` which will open the handle anyway.
But also I understood that optimising identical images will do wasteful operations in parallel, so an even better solution would be to have static tmp file name (like `cached` with `.tmp` extension) and exclusively lock that file, check if it still has zero size, run the block, write data, rename to `cached`.

gpakosz commented:
I don't really understand what you don't like with the current `tmp_name` private method then using `FileUtils.mv`, beside the aesthetics.

Performance wise, `FileUtils.mv` defers to `File.rename` is both `src` and `dest` are on the same filesystem. Otherwise it copies the file but well if it's not the case already, this copy can be kept low level ala Java NIO.

Having a static `@cache_dir / 'cached.tmp'` temp file and locking with a mutex is really a waste imho: there's no reason to impose such lock contention.

When caching is enabled, it seems natural to **move** the optimized file to `@cache_dir` and in fact now that I think about it the ultimate choice is to use the basename of the optimized file since it's already a unique temporary file name:

        tmp = @cache_dir / File.basename(optimized)
        FileUtils.mv(optimized, tmp, :force => true)
        File.rename(tmp, cached)

toy commented:
When checking the complete process, I've found a problem related to cache:
Without cache the `optimize_image` method returns a wrapped temp file path which in `optimize_image!` replaces the original file, so the temp file is removed as it is not needed anymore. With cache the returned file is the path in cache, so the cached file will be removed. Just try to use `image_optim` binary on a folder with cache enabled and check the cache directory. I can suggest creating a class for cache path which will override the `replace` method which will do the same but will not remove the file from cache.

About locking I may have been unclear:
The cache directory will be shared by multiple parallel optimisations, it can also be shared by multiple processes doing optimisation, so if an identical image is optimised in parallel, tools will be wasting cycles as they will overwrite the cache file with same result.
That is why I proposed locking (not mutex, but `flock`) which will also make creating unique name unneeded. `flock` is not slow (tens of thousands open'n'flock per second) especially compared to time to optimise an image.
So using something like `tmp = @cache_dir / "#{digest}.tmp"`, open it with `File::CREAT | File::WRONLY` and `flock(File::LOCK_EX)`, repeat checks `file?` and `symlink?` as the cache file may have already been written and only then yield and eventually move result and rename or create a symlink.

`FileUtils.mv` is indeed good performance wise, I was already thinking that the way `ImagePath#replace` copies and removes file is wasteful when source and destination are on the same volume.

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,64 @@
+      optimized = yield
+
+      digest = digest(original)
+      cached = @cache_dir / digest
```
`cached` is already assigned previously, also `digest` on previous line is also not needed

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,64 @@
+
+      cached = @cache_dir / digest(original)
+      return cached if File.file?(cached)
+      return if File.symlink?(cached)
```
`cached.symlink?`

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,64 @@
+      return yield unless @cache_dir
+
+      cached = @cache_dir / digest(original)
+      return cached if File.file?(cached)
```
`cached.file?`

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,64 @@
+      if optimized
+        tmp = @cache_dir / File.basename(optimized)
+        FileUtils.mv(optimized, tmp, :force => true)
+        File.rename(tmp, cached)
```
`tmp.rename(cached)`

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,64 @@
+
+      if optimized
+        tmp = @cache_dir / File.basename(optimized)
+        FileUtils.mv(optimized, tmp, :force => true)
```
`:force => true` only affects removal of original when copying, so should not be needed

gpakosz commented:
Are there expectations about `#optimize_image` returning a wrapped temp file? If so, shouldn't it return one no matter what, which brings us back to generating a temp file name without opening a file handle.

toy commented:
The only place in `image_optim` expecting a temp file is `#optimize_image!` as it removes the path returned by `#optimize_image`, but there can be a problem in external usage of `#optimize_image`.
Though always copying to temp file is unneeded work for methods like `#optimize_image_data` as it just reads data from file (GC takes care of removing the temp file).
Besides returning a special class with custom `#replace` (which will ensure only proper internal use), I've thought about returning a hardlink or a symlink to cache path or some more complicated solutions. What do you think about always returning a temp file, but adding an option like `allow_cache_path` to `#optimize_image` and `Cache#fetch` to return a direct cache path or is it exposing too much internal implementation?

gpakosz commented:
> The cache directory will be shared by multiple parallel optimisations, it can also be shared by multiple processes doing optimisation, so if an identical image is optimised in parallel, tools will be wasting cycles as they will overwrite the cache file with same result.

How common is this situation? While it can theoretically happen, I have difficulties assuming it's not a degraded case.

I was about to go with subclassing `ImagePath::Optimized` but I find `ImageOptim::ImagePath::Optimized` being a proxy for `ImageOptim::ImagePath` a bit convoluted.

gpakosz commented:
About always returning a temp file, I think I prefer keeping the `CachePath` with a non-destructive `#replace()` method. Client code that scavenges cache by moving files away will have to adapt.

gpakosz commented:
Alright, I implemented what we discussed and rebased everything onto the master branch.

toy commented:
I agree, optimising identical images in parallel probably happens rarely and can be done much later if there will be a need of that. But then I urge you to use a temp file path for `tmp` in `#fetch`. Even if `optimized.basename` should be a name of a temp file, creating a temp file is not an expensive operation compared to optimisation itself.

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,65 @@
+
+      cached = @cache_dir / digest(original, original.format)
+      return cached if cached.file?
+      return if cached.symlink?
```
Better to first check `symlink?` then `file?` as symlink to file will get true for `file?`

gpakosz replied:
Yeah I can swap the two. But we symlink to `@cache_dir` anyways

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,65 @@
+
+      if optimized
+        tmp = @cache_dir / optimized.basename
+        FileUtils.mv(optimized, tmp)
```
Fails if `@cache_dir` is not yet created, moving after mkpath will solve it

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,65 @@
+    def fetch(original)
+      return yield unless @cache_dir
+
+      cached = @cache_dir / digest(original, original.format)
```
`format` is now `image_format`

gpakosz replied:
yeah the rebase was painful, fixing that first before doing anything else

gpakosz replied:
and now that it's more verbose rubocop complains about lines > 80 character in tests :(

toy replied:
Sorry for conflicting changes )
But if you use `ImageOptim::Path` for `@cache_dir`, you will have `image_format` for both `original` and `cached`, so only one parameter needed.

toy replied:
About 80 character limit – I'm already getting tired of it, so I'm okay to change it to for example 120.

gpakosz replied:
well in fact it's the first commit of the series that was annoying, then I found a way to write shorter lines

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,65 @@
+        FileUtils.ln_sf(@cache_dir, cached_)
+        CachePath.convert(cached)
+      else
+        FileUtils.ln_sf(@cache_dir, cached)
```
`cached.dirname.mkpath` is also needed before this

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,65 @@
+        FileUtils.mv(optimized, tmp)
+        cached.dirname.mkpath
+        tmp.rename(cached)
+        cached_ = @cache_dir / digest(cached, original.format)
```
Confusing variable name, maybe extract a method for marking unoptimizable images?

gpakosz replied:
I avoided using `cached_` and reused `cached` instead, I added a comment to explain the purpose of symlinking. Tell me how it looks to you

toy commented on lib/image_optim/cache_path.rb:
```diff
@@ -0,0 +1,22 @@
+
+    # Returns path if it is already an instance of this class otherwise new
+    # instance
+    def self.convert(path)
```
Method from superclass is already doing exactly the same

gpakosz replied:
Removed

toy commented on lib/image_optim/cache_path.rb:
```diff
@@ -0,0 +1,22 @@
+    def replace(dst)
+      dst = self.class.new(dst)
+      dst.temp_path(dst.dirname) do |temp|
+        dst.copy(temp)
```
I think this and next line should be
```ruby
copy(temp)
dst.copy_metadata(temp)
```

gpakosz replied:
Ah yes, something I overlooked in the rebase as well

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,65 @@
+      return yield unless @cache_dir
+
+      cached = @cache_dir / digest(original, original.format)
+      return cached if cached.file?
```
`cached` must be converted to `CachePath`

gpakosz commented:
>But then I urge you to use a temp file path for tmp in #fetch. Even if optimized.basename should be a name of a temp file, creating a temp file is not an expensive operation compared to optimisation itself.

I don't understand how taking `optimized.basename` could clash with anything in fact, care to elaborate?

toy commented:
The important thing about temp file is that underlying calls ensure that it can't clash. Different implementation of Tmpname or different usage of caching can cause weird bugs.

gpakosz commented:
Ok this batch of commits should give you a better impression. Last time I did my tests before rebasing and breaking everything :/

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,72 @@
+      digest = digest(original, original.image_format)
+      cached = @cache_dir / digest
+      return CachePath.convert(cached) if cached.file?
+      return if cached.symlink? && File.readlink(cached) == @cache_dir.to_s
```
Not good if cache dir is moved or if it is specified differently ('cache', './cache', '/full/path/to/cache').
This made me think about what is it better to link to. I understand that we should not care about link destination, currently it will be most probably broken if cache dir is relative. It can also be a fake location like '-' to make it very small or file digest to make it descriptive. What do you think?

gpakosz replied:
Totally solvable with `@cache_dir = FSPath.new(image_optim.cache_dir).realpath` in constructor?

gpakosz replied:
`'-'` or any ASCII file name will do right? No possible SHA1 can end up being a intelligible word nor `-`

gpakosz replied:
You're right. In fact I don't believe symlinking is a good idea anymore. Now, I'm checking whether cached size is empty to decide whether it corresponds to an already optimized file. This solves moving the cache directory.

gpakosz commented:
As mentioned in the comment above, now, I'm checking whether cached size is empty to decide whether it corresponds to an already optimized file. This solves moving the cache directory.

toy commented:
Do you mean that you wanted to create a symlink to `da39a3ee5e6b4b0d3255bfef95601890afd80709`? That is a sha1 of an empty string as `/dev/null` can't be read.
Currently windows is not supported by `image_optim`, so this should not be a reason, but I'm also not sure about using symlinks and maybe it is not hard to add support of windows, so better to not make it harder.

Overall there should be no way for an empty file to appear when creating an entry for a succeeded optimisation, so using empty files for unoptimisable images should not be a problem. I had ideas like adding an extension to those or putting them inside a sub-folder, but this will bring more code and no much gain.

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,72 @@
+      digest = digest(original, original.image_format)
+      cached = @cache_dir / digest
+      return if cached.file? && cached.zero?
+      return CachePath.convert(cached) if cached.file?
```
Can be `return cached.size? ? CachePath.convert(cached) : nil if cached.file?` to reduce number of file checks

gpakosz replied:
Don't you think it's a little less readable?

gpakosz commented:
Yes I hesitated with creating a symlink to `/path/to/cache_dir/da/39a3ee5e6b4b0d3255bfef95601890afd80709` but in fact that's not needed. As you noticed, the situation where `image_optim` would produce zero sized files in cache should be when caching already optimized images.

Do you think we're good?

toy commented on lib/image_optim/runner/option_parser.rb:
```diff
@@ -204,6 +204,19 @@ def wrap_regex(width)
   end
 
   op.separator nil
+  op.separator '  Caching:'
```
Please move this section higher, after main options and before disabling workers

toy commented on lib/image_optim/cache.rb:
```diff
@@ -0,0 +1,71 @@
+
+      digest = digest(original, original.image_format)
+      cached = @cache_dir / digest
+      return cached.size? ? CachePath.convert(cached) : nil if cached.file?
```
It can even be `return cached.size? && CachePath.convert(cached) if cached.file?` as `size?` returns `nil` when it doesn't return a number

toy commented:
Please ignore AppVeyor failing, I will try to test how far is `image_optim` from working on windows.

gpakosz commented:
Yep will do. I noticed it's not configured yet.

toy commented:
Few last points about tests:
- Please add a test for `CachePath#replace` (you can adapt one for [Path#replace](https://github.com/toy/image_optim/blob/master/spec/image_optim/path_spec.rb#L55))
- Test groups for "when cache is enabled" with and without worker digests are very similar, maybe a shared test group?
- Better to check that only `CachePath` or `nil` can be returned and never just `Path`
- Marking cached image as already optimised is not tested
- And behaviour when zero size cache file exists

gpakosz commented:
I'll look into it. I don't know RSpec well though so I'm not sure about what shared test groups are and how to use them

toy commented:
Shared examples in short:
```ruby
shared_examples "does something" do
  it … do
  end
end

describe … do
  let …
  before …
  it_behaves_like "does something"
end
```

gpakosz commented:
Yep I finally found my through it this afternoon :)

toy commented:
Hi Gregory, can you please rebase the branch on master, sqash-split to one or more atomic/logical commits and check if it works on appveyor?

toy commented:
I'll extract capabilities checkers for easy reusing, so failing windows checks can be ignored

toy commented:
Have a look at `CapabilityCheckHelpers`, `any_file_modes_allowed?` and `inodes_supported?` in `path_spec.rb`

gpakosz commented:
I updated `cache_path_spec.rb` from `path_spec.rb`.

gpakosz commented:
What do you prefer between keeping the 10 commits from the `cache` branch, and squashing everything into a single commit?

For the record, tests and rubocop are all good in every of the 10 commits. I let you decide how you prefer history to look like

toy commented:
Those are development commits, so better 1 commit unless you can and would like to split them into several logical separate ones.
Also please add the link to this PR in changelog entry.

gpakosz commented:
Is it important to link the PR? I linked #83.

By the way, why are you using `[#XX](https://github.com/toy/image_optim/issues/XX)` in `CHANGELOG.markdown`? Using `#XX` is enough, GitHub automatically creates links.

toy commented:
Already not, as important, as you've just linked them by mentioning ;)

I already wanted to check it for some time and thank you for pushing me to do it :), but unfortunately it works only in comments/descriptions and some other stuff, but not in markdown files:
https://github.com/toy/image_optim/blob/a4c2767f7b088f9ca091fd33ba04f9e772f805a7/CHANGELOG.markdown
https://github.com/toy/image_optim/blob/62c85009e3eea006ad47bd5cf0d032d3215ece5f/CHANGELOG.markdown
https://github.com/toy/image_optim/blob/657ed3f1d634608a4b02ef54eeaa50f92b029eb0/CHANGELOG.markdown

gpakosz commented:
Oh indeed you're right

toy commented:
Thank you for finishing this long PR, I'll release a new version soon

gpakosz commented:
You're welcome. Thanks for the review and the suggestions. That was nice.
